mitiq/cdr/cdr_execution.py

```diff
@@ -0,0 +1,208 @@
+# Copyright (C) 2021 Unitary Fund
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+""" Functions for execution of CDR mitigation on circuit of interest"""
+
+import numpy as np
+
+from cirq.circuits import Circuit
+
+from typing import List, Union, Callable, Optional
+
+from clifford_training_data import generate_training_circuits
+
+from data_regression import (
+    scale_noise_in_circuits,
+    execute_training_circuits,
+    execute_circuit_of_interest,
+    construct_training_data_floats,
+    construct_circuit_data_floats,
+    find_optimal_parameters,
+    wrapper_fit_func,
+    calculate_observable,
+)
+
+
+def execute_with_CDR(
+    circuit: Circuit,
+    executor: Callable[[Circuit], dict],
+    simulator: Callable[[Circuit], Union[dict, np.ndarray]],
+    observable: Union[List[np.ndarray], np.ndarray],
+    num_training_circuits: int,
+    fraction_non_clifford: float,
+    scale_noise: Optional[Callable[[Circuit, float], Circuit]] = None,
+    noise_scaling_factors: Optional[Union[float, List[float]]] = None,
```

Comment:
How common is it to have just a float? Can we just force this to always be a list?

Reply:
Yes I think thats fine.

Reply:
Great. That's consistent with setting scale factors in `Factory`s which always take a list, and makes the function easier for you to implement.